### ISSUE TITLE
Segmentation fault when build with DTLS=1

### ISSUE THREAD

**Issue by ieei0214** [eclipse/wakaama#134]
I'm not familiar with the new build system.  Does anyone have the same problem?

ricky@ricky-VirtualBox:~/lwm2m/wakaama/bc$ cmake -DDTLS=1 ../examples/client/
-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- Found Git: /usr/bin/git (found version "1.9.1")
-- Configuring done
Segmentation fault (core dumped)

**Comment by ieei0214** [comment]
It's weird.  I found that if I let cmake do "autoreconf -i" and "configure" for tinydtls, it will get "Segmentation fault".

If I do "autoreconf -i" and "configure" in tinydtls directory manually, it works fine.

I mark those lines in tinydtls.cmake.  Anyone get the same issue?

#if (NOT EXISTS ${TINYDTLS_SOURCES_GENERATED})
#    include(ExternalProject)
#    ExternalProject_Add(external_tinydtls
#        SOURCE_DIR  "${TINYDTLS_SOURCES_DIR}"
#        DOWNLOAD_COMMAND ""
#        UPDATE_COMMAND ""
#        BUILD_COMMAND ""
#        INSTALL_COMMAND ""
#        CONFIGURE_COMMAND "${TINYDTLS_SOURCES_DIR}/configure"
#        BUILD_IN_SOURCE 1
#        LOG_DOWNLOAD 1
#        LOG_CONFIGURE 1
#        # Make the submodule_update target a dependency.
#        DEPENDS submodule_update
#        )
#
#    ExternalProject_Add_Step(external_tinydtls configure2
#       COMMAND "autoreconf" "-i"
#       ALWAYS 1
#       WORKING_DIRECTORY "${TINYDTLS_SOURCES_DIR}"
#       DEPENDERS configure
#       DEPENDEES download
#    )
#
#    # Let cmake know that it needs to execute the external_tinydtls target to generate those files.
#    add_custom_command(OUTPUT ${TINYDTLS_SOURCES_GENERATED} COMMAND ""  DEPENDS external_tinydtls)
#endif()

**Comment by dnav** [comment]
I just did this (in /tmp):
```git clone https://github.com/eclipse/wakaama.git```
```git submodule init```
```git submodule update```
```mkdir build ; cd build```
```cmake -DDTLS=1 ../examples/client/```
and everything went fine.
Note that the autoreconf step occurs when compiling not during cmake.

Regards,

**Comment by dnav** [comment]
PS:
cmake version 3.4.1 on a 64-bit Fedora 23

**Comment by ieei0214** [comment]
I remove those lines in tinydtls.cmake and it can work fine.

I don't know why it can't work so I do "autoreconf -i" and "configure" by myself.

>find_package(Git REQUIRED)                                       
>add_custom_command(OUTPUT ${TINYDTLS_SOURCES} COMMAND ${GIT_EXECU
>add_custom_target(submodule_update SOURCES ${TINYDTLS_SOURCES})  

>if (NOT EXISTS ${TINYDTLS_SOURCES_GENERATED})
>    include(ExternalProject)
>    ExternalProject_Add(external_tinydtls
>        SOURCE_DIR  "${TINYDTLS_SOURCES_DIR}"
>        DOWNLOAD_COMMAND ""
>        UPDATE_COMMAND ""
>        BUILD_COMMAND ""
>        INSTALL_COMMAND ""
>        CONFIGURE_COMMAND "${TINYDTLS_SOURCES_DIR}/configure"
>        BUILD_IN_SOURCE 1
>        LOG_DOWNLOAD 1
>        LOG_CONFIGURE 1
>        # Make the submodule_update target a dependency.
>        DEPENDS submodule_update
>        )
>
>    ExternalProject_Add_Step(external_tinydtls configure2
>       COMMAND "autoreconf" "-i"
>       ALWAYS 1
>       WORKING_DIRECTORY "${TINYDTLS_SOURCES_DIR}"
>       DEPENDERS configure
>       DEPENDEES download
>    )
>
>    # Let cmake know that it needs to execute the external_tinydtls target to generate those files.
>    add_custom_command(OUTPUT ${TINYDTLS_SOURCES_GENERATED} COMMAND ""  DEPENDS external_tinydtls)
>endif()

**Comment by ieei0214** [comment]
cmake version 2.8.12.2

Update to cmake version 3.5.1 and then test again.  It works fine.

**Comment by davidgraeff** [comment]
Cmake 2.8.12 is three years old. But some long time support operating systems still ship that (and don't update their buggy toolchains anymore) unfortunately.
This can be solved by allowing ExternalProject_Add only if cmake is Version 3.0 or newer and printing a warning to the user otherwise.

**Comment by dnav** [comment]
Let's keep things simple. Requiring a minimum version 2.8.3 does not seem exaggerated to me.
Proposed fix: #135.

**Comment by ieei0214** [comment]
I agree.  But right now it's already min required 2.8.3 in CMakeLists.txt.

Here is my test.
cmake version 2.8.12.2  =>  fail.
cmake version 3.5.1  => ok.

**Comment by dnav** [comment]
My bad, I read your version as 2.8.1.2

**Comment by dnav** [comment]
I bumped it to 3.0

**Comment by davidgraeff** [comment]
I propose something like
```
if(CMAKE_MAJOR_VERSION LESS 3)
 message("Your cmake version is very old and will propably crash while configuring tinydtls!")
endif()
```

**Comment by davidgraeff** [comment]
Of course if would be great if cmake 3 is the minimum requirement everywhere.

**Comment by dnav** [comment]
Fixed by #135
